### ISSUE TITLE
Investigate whether this plugin is still required

### ISSUE THREAD

**Issue by nmfisher** [rmawatson/flutter_isolate#128]
Per https://github.com/flutter/flutter/issues/13937#issuecomment-1258766459, Flutter now seems to have support for running plugin code in background isolates, which would render this plugin redundant.

This is apparently merged into Flutter master (not yet stable as of 12 November 2022) - I don't have much free time so would appreciate someone investigating further to see if this is correct.

**Comment by rlueders** [comment]
I have been using your plugin to call Flutter UI functions from an isolate. I will check the sample in the link you provided to see if it offers a solution.

**Comment by rlueders** [comment]
It appears that changes from 13937 do not allow for running dart.ui code from a standard Isolate. Consequently, this plugin does indeed provide additional capability outside of standard flutter isolates and is not completely redundant.   

After registering a background isolate with a RootIsolateToken as described in 13937 sample code, the isolate will still throw a "UI Actions are only available on root isolate" exception any time that dart.ui code is called from a standard isolate. No special handling is required when calling dart.ui code from a FlutterIsolate (this plugin).

I tried enabling the new impeller render on IOS to see if there was any difference in behavior, but the same exception is thrown for any calls to dart.ui code from a standard Isolate.  Surprisingly, the calls to dart.ui.picture.toImage() (with impeller enabled) from  FlutterIsolate (this plugin) result in "Exception: operation failed" exception. As such, this plugin is currently the only way to run dart.ui code from an isolate, but it looks like that might be breaking with the new impeller renderer unless it is fixed before the stable release of impeller.

**Comment by nmfisher** [comment]
Thank you @rlueders, this is excellent. Let me review further and see what's needed for Impeller.

**Comment by rlueders** [comment]
> Thank you @rlueders, this is excellent. Let me review further and see what's needed for Impeller.

Did you ever get a chance to look into this? looks like the same behavior on the stable release of Flutter 3.7.

**Comment by rlueders** [comment]
Here are a few other related issues requesting dart.ui from isolates: 

- [13343 ](https://github.com/flutter/flutter/issues/13343)
- [75755](https://github.com/flutter/flutter/issues/75755)
- [53985](https://github.com/flutter/flutter/issues/53985)
- [30604](https://github.com/flutter/flutter/issues/30604)

The reason the Flutter engine does not allow dart.ui from background isolate is discussed in  [13343](https://github.com/flutter/flutter/issues/13343#issue-279200892) by @chinmaygarde where he talks about dart.ui having some unsafe calls related to frame scheduling.  The majority of the function in dart.ui are safe to call from a background isolates.  His proposal to "Design a dart:ui lite variant of bindings that are safely accessible by secondary isolates." makes sense, but since this has been open for 5 years, I doubt it will be prioritized any time soon.

**Comment by jamesncl** [comment]
Also isolates still do not allow access the root bundle, so you can't load asset files. This bit me when attempting to remove `flutter_isolate` from my project, because I was trying to initialise a timezone package using a timezone database file in my assets. So for me, this package is still useful! This is mentioned in [a comment in the official isolate example code](https://github.com/flutter/flutter/blob/c302232e3a995243d9f38daeeca7757d3b6758f9/examples/layers/services/isolate.dart#L130):

```
//  ... spawned isolates do not have access to the root bundle
```

However I do think the docs need updating. [https://pub.dev/packages/isolate_handler](https://pub.dev/packages/isolate_handler) currently says the following which is no longer true for 3.7 +

> Effortless isolates abstraction layer with support for inter-isolate communication. These isolates (unlike the standard ones in Flutter) can call platform plugins.

**Comment by nmfisher** [comment]
That's a good point - annoyingly, I think flutter forces all assets loading to occur on the main thread (flutter_isolate circumvents this by spawning a new FlutterEngine which I assume itself starts a new thread(s)).

You're also right that a documentation update is probably needed.

**Comment by greensopinion** [comment]
> It appears that changes from 13937 do not allow for running dart.ui code from a standard Isolate. Consequently, this plugin does indeed provide additional capability outside of standard flutter isolates and is not completely redundant.

I'm interested in this capability to improve the frame rate of [vector_map_tiles](https://pub.dev/packages/vector_map_tiles) which renders map tiles from vector data. Currently the plugin suffers from jank.

> I tried enabling the new impeller render on IOS to see if there was any difference in behavior, but the same exception is thrown for any calls to dart.ui code from a standard Isolate. Surprisingly, the calls to dart.ui.picture.toImage() (with impeller enabled) from FlutterIsolate (this plugin) result in "Exception: operation failed" exception. As such, this plugin is currently the only way to run dart.ui code from an isolate, but it looks like that might be breaking with the new impeller renderer unless it is fixed before the stable release of impeller.

Has anyone raised this issue with the Flutter team? They may not be aware of this use-case.

**Comment by greensopinion** [comment]
Thanks for the insights, I appreciate it. 

From what I can gather, the main difference between those issues and this case is that here we have an existing capability that is no longer working, whereas the others are for adding a capability that doesn't exist today.

**Comment by rlueders** [comment]
FYI this issue (calling dart.ui functions from a Flutter Isolate) was resolved in one of the recent Flutter releases.

**Comment by nmfisher** [comment]
Thanks @rlueders - I guess that's one more argument in favour of deprecating this plugin.

**Comment by greensopinion** [comment]
> this issue (calling dart.ui functions from a Flutter Isolate) was resolved in one of the recent Flutter releases.

Any idea where we can read about this? I haven't found it.

**Comment by rlueders** [comment]
> 

No, it is an argument in favor of keeping the plugin.  I can see how my previous reply was confusing. Let me rephrase.  Calling dart.ui from  a standard platform Isolate has never worked an still doesn't work.  However, calling dart.ui from a Flutter Isolate (this plugin) now works when using the latest version of Flutter/Dart. It was actually fixed in this [dart change](https://dart-review.googlesource.com/c/sdk/+/292320).

**Comment by nmfisher** [comment]
Understood, thanks for clarifying.
